src/proofnetwork/xjdf/data/part.cpp

```diff
@@ -0,0 +1,168 @@
+/* Copyright 2019, OpenSoft Inc.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are permitted
+ * provided that the following conditions are met:
+ *
+ *     * Redistributions of source code must retain the above copyright notice, this list of
+ * conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above copyright notice, this list of
+ * conditions and the following disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *     * Neither the name of OpenSoft Inc. nor the names of its contributors may be used to endorse
+ * or promote products derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
+ * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+#include "proofnetwork/xjdf/data/part.h"
+
+#include "proofnetwork/xjdf/data/cutblock.h"
+#include "proofnetwork/xjdf/data/cuttingparams.h"
+#include "proofnetwork/xjdf/data/product.h"
+#include "proofnetwork/xjdf/data/productlist.h"
+#include "proofnetwork/xjdf/data/resourceset.h"
+#include "proofnetwork/xjdf/data/xjdfabstractnode_p.h"
+#include "proofnetwork/xjdf/data/xjdfdocument.h"
+
+namespace Proof {
+namespace XJdf {
+class PartPrivate : public XJdfAbstractNodePrivate
```

Comment:
I'm completely lost about what we should do with parts. Two main questions:
1. What combinations of various partitions are allowed (there are 15 or so of them according to documentation) and should we verify all these combinations or not (i.e. should we nullify something if incompatible attribute was set)?
2. Should we store here DTOs (and return them in getters) or strings will suffice. DTOs is a common way we use in most cases and it gives us easier access to all these relations. However, in case of jdf we always have the full tree and have all these traversing mechanisms, so probably we can skip caches here completely (which will give us some advantages like eliminating issues with duplicated IDs) and use strings. It will slow down traversing on huge jdfs (not sure actually how slow it will be, probably insignificant comparing to downloading this jdf from network), but will give us more freedom. I'm not talking about replacing everything with string of course, you still need to keep list of cutblocks in cuttingparams or component in resource. I'm looking right now on Part and on anything that replaced Links from jdf (if such mechanism still exists in xjdf).